### PR TITLE
Ops for 2024.0 release

### PR DESCRIPTION
**Pull request by suryasidd**
### Details:
 - Added operators for 2023.3 release
 - Includes ops required for enabling customer model lightglue. 
 - Also added ops for Lora fine tuning

### Tickets:
 - https://jira.devtools.intel.com/browse/CVS-116701

### CODE REVIEW

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend.py [hunk 1]
```diff
@@ -111,6 +114,43 @@ def _call(*args):
...
+):
+
+    # Compute the gradient of the input tensor
+    grad_input = torch.nn.functional.conv_transpose2d(
+        grad_output, weight, stride=stride, padding=padding, dilation=dilation, groups=groups
+    )
+
+    # Compute the gradient of the weight tensor
+    grad_weight = torch.nn.functional.conv_transpose2d(
+        #input, weight.transpose(0,1), stride=stride, padding=padding, dilation=dilation, groups=groups
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend.py:140]: looks like some debug comment, please remove?

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend_utils.py [hunk 2]
```diff
@@ -27,8 +27,16 @@ def _get_device(options) -> Optional[Any]:
...
+    decompositions = []
+    if options is not None and "decompositions" in options:
+        decompositions = options["decompositions"]
+    return decompositions
+
+
 def _is_cache_dir_in_config(options) -> Optional[Any]:
-    if "config" in options:
+    if options is not None and "config" in options:
         cfg = options["config"]
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend_utils.py:40]: should here be also check that config is not None?

File: src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend_utils.py [hunk 3]
```diff
@@ -57,4 +65,4 @@ def _get_model_caching(options) -> Optional[Any]:
 
 def _get_config(options) -> Optional[Any]:
     if options is not None and "config" in options:
-        return options["config"]
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/torchdynamo/backend_utils.py:60]: please add return value for case if config is not in options

File: src/frontends/pytorch/src/op/min_max.cpp [hunk 4]
```diff
@@ -50,6 +51,55 @@ OutputVector translate_max(const NodeContext& context) {
...
+    auto k = context.mark_node(std::make_shared<v0::Constant>(element::i32, Shape{}, 1));
+    auto topk = std::make_shared<v3::TopK>(x, k, axis_const, v3::TopK::Mode::MAX, v3::TopK::SortType::NONE);
+    auto indicies = context.mark_node(std::make_shared<v0::Convert>(topk->output(1), element::i64));
+    if (!keepdims) {
+        indicies = std::make_shared<v0::Squeeze>(indicies, axes_node);
+    }
+    return {values, indicies};
+};
+
+OutputVector translate_max_dim_fx(const NodeContext& context) {
```
Comment (eaidova) [inline src/frontends/pytorch/src/op/min_max.cpp:77]: I do not see the difference except packing output between 2 translators, can it reuse one inside another?

File: src/frontends/pytorch/src/op/var_mean.cpp [hunk 5]
```diff
@@ -12,6 +12,8 @@
 #include "openvino/op/shape_of.hpp"
 #include "openvino/op/sqrt.hpp"
 #include "openvino/op/subtract.hpp"
+#include "openvino/op/unsqueeze.hpp"
```
Comment (eaidova) [inline src/frontends/pytorch/src/op/var_mean.cpp:15]: I do not see unsqueeze usage in this file? Does these headers added accidentally?

File: src/frontends/pytorch/src/op/var_mean.cpp [hunk 6]
```diff
@@ -76,6 +78,43 @@ OutputVector translate_var_mean(const NodeContext& context) {
...
 
+OutputVector translate_var_mean_fx(const NodeContext& context) {
+    num_inputs_check(context, 1, 4);
+    auto data = context.get_input(0);
+    bool unbiased = false;
+    auto num_elements = numel(context, data);
+    std::shared_ptr<ov::Node> mean, t_mean;
+    ov::Output<ov::Node> axes;
+
+    axes = context.get_input(1);
```
Comment (eaidova) [inline src/frontends/pytorch/src/op/var_mean.cpp:89]: the check of number of inputs does not look correct. Above you check that at least 1 input provided, but here you use the second input without any checks that context contains it

File: src/frontends/pytorch/src/op/var_mean.cpp [hunk 7]
```diff
@@ -76,6 +78,43 @@ OutputVector translate_var_mean(const NodeContext& context) {
...
+    auto axis_0 = context.mark_node(v0::Constant::create(element::i32, Shape{}, {0}));
+    mean = context.mark_node(std::make_shared<v1::ReduceMean>(data, axes, true));
+    t_mean = context.mark_node(std::make_shared<v1::ReduceMean>(data, axes, true));
+    auto reduced_dims = context.mark_node(std::make_shared<v3::ShapeOf>(data, element::i32));
+    auto zero = context.mark_node(v0::Constant::create(element::i32, Shape{}, {0}));
+    reduced_dims = context.mark_node(std::make_shared<v8::Gather>(reduced_dims, axes, zero));
+    num_elements = context.mark_node(std::make_shared<v1::ReduceProd>(reduced_dims, zero, false));
+
+    auto sub_v = context.mark_node(std::make_shared<v1::Subtract>(data, t_mean));
```
Comment (eaidova) [inline src/frontends/pytorch/src/op/var_mean.cpp:98]: is it possible to reuse some code between translators? Looks very similar to ts translator, except output replacing and first input check

File: src/frontends/pytorch/src/op_table.cpp [hunk 8]
```diff
@@ -606,12 +613,17 @@ const std::map<std::string, CreatorFunction> get_supported_ops_ts() {
 const std::map<std::string, CreatorFunction> get_supported_ops_fx() {
     return {
         {"aten._adaptive_avg_pool2d.default", op::translate_1to1_match_2_inputs<opset10::AdaptiveAvgPool>},
+        {"aten._fake_quantize_per_tensor_affine_cachemask_tensor_qparams.default",
+         op::translate_fake_quantize_per_tensor_affine_fx},
+        {"aten::_log_softmax.default", op::translate_log_softmax},
```
Comment (eaidova) [inline src/frontends/pytorch/src/op_table.cpp:618]: Probably you mean:
```suggestion
        {"aten._log_softmax.default", op::translate_log_softmax},
```

**Comment by p-durandin** [comment]
Pls fix windows build

**Comment by suryasidd** [comment]
We don't need to merge this PR for 2023.3. Removing the labels

File: src/frontends/pytorch/src/op/log.cpp [hunk 9]
```diff
@@ -29,6 +30,15 @@ OutputVector translate_log(const NodeContext& context) {
     return {log};
 };
 
+OutputVector translate_log_sigmoid(const NodeContext& context) {
+    num_inputs_check(context, 1, 1);
+    auto x = context.get_input(0);
+    auto sigmoid = context.mark_node(std::make_shared<v0::Sigmoid>(x));
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/log.cpp:36]: Should the f32 conversion happen after Sigmoid? Is this the expected behavior for int-type values?

File: src/frontends/pytorch/src/op/min_max.cpp [hunk 10]
```diff
@@ -50,6 +51,55 @@ OutputVector translate_max(const NodeContext& context) {
...
+
+    bool keepdims = false;
+    if (!context.input_is_none(2)) {
+        keepdims = context.const_input<bool>(2);
+    }
+
+    auto values = context.mark_node(std::make_shared<v1::ReduceMax>(x, axes_node, keepdims));
+    auto k = context.mark_node(std::make_shared<v0::Constant>(element::i32, Shape{}, 1));
+    auto topk = std::make_shared<v3::TopK>(x, k, axis_const, v3::TopK::Mode::MAX, v3::TopK::SortType::NONE);
+    auto indicies = context.mark_node(std::make_shared<v0::Convert>(topk->output(1), element::i64));
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/min_max.cpp:70]: Minor grammar change
```suggestion
    auto indices = context.mark_node(std::make_shared<v0::Convert>(topk->output(1), element::i64));
```

File: src/frontends/pytorch/src/op/min_max.cpp [hunk 11]
```diff
@@ -50,6 +51,55 @@ OutputVector translate_max(const NodeContext& context) {
...
+    if (!keepdims) {
+        indicies = std::make_shared<v0::Squeeze>(indicies, axes_node);
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/min_max.cpp:72]:
```suggestion
        indices = std::make_shared<v0::Squeeze>(indices, axes_node);
```

File: src/frontends/pytorch/src/op/min_max.cpp [hunk 12]
```diff
@@ -50,6 +51,55 @@ OutputVector translate_max(const NodeContext& context) {
...
+    }
+    return {values, indicies};
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/min_max.cpp:74]:
```suggestion
    return {values, indices};
```

File: src/frontends/pytorch/src/op/split.cpp [hunk 13]
```diff
@@ -35,6 +35,23 @@ OutputVector translate_chunk_fx(const NodeContext& context) {
     return {context.mark_node(make_list_construct(chunk->outputs()))};
 }
 
+OutputVector translate_unbind_int_fx(const NodeContext& context) {
+    num_inputs_check(context, 2, 3);
+    auto input = context.get_input(0);
+    auto dim = context.get_input(1);
+    auto dim_val = context.const_input<int>(1);
+    auto shape = context.get_input(0).get_shape();
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/split.cpp:43]:
```suggestion
    auto shape = input.get_shape();
```

File: src/frontends/pytorch/src/op/split.cpp [hunk 14]
```diff
@@ -35,6 +35,23 @@ OutputVector translate_chunk_fx(const NodeContext& context) {
...
+
+    if (dim_val < 0) {
+        dim_val = static_cast<int>(shape.size()) + dim_val;
+    }
+
+    auto num_splits = static_cast<int>(shape[dim_val]);
+    auto chunk = context.mark_node(std::make_shared<v1::Split>(context.get_input(0), dim, num_splits));
```
Comment (PiotrKrzem) [inline src/frontends/pytorch/src/op/split.cpp:50]:
```suggestion
    auto chunk = context.mark_node(std::make_shared<v1::Split>(input), dim, num_splits));
```

**Review by PiotrKrzem — CHANGES_REQUESTED** [review]

**Comment by suryasidd** [comment]
Closing this, opened a new one. https://github.com/openvinotoolkit/openvino/pull/22745
